File: app/utils/ollama.ts
import { Command } from '@tauri-apps/plugin-shell'

export const ollama = { isReady: false, port: '11434' }

function sleep(ms = 500) {
  return new Promise(resolve => setTimeout(resolve, ms))
}

export async function runOllamaServe() {
  if (ollama.isReady)
    return

  Command.sidecar('binaries/ollama', [
    'serve',
  ]).execute()

  await sleep(500)
  ollama.isReady = true
}

File: src-tauri/src/core/ollama.rs
use super::GlobalState;
use std::sync::Mutex;
use std::thread;
use std::time::Duration;
use tauri::{AppHandle, Manager};
use tauri_plugin_shell::ShellExt;

pub fn run_ollama_serve(app: &AppHandle) {
    let sidecar_command = app.shell().sidecar("ollama").unwrap().args(["serve"]);
    let (_, child) = sidecar_command.spawn().expect("Failed to spawn sidecar");
    thread::sleep(Duration::from_millis(500));
    let state = app.state::<Mutex<GlobalState>>();
    let mut state = state.lock().unwrap();
    state.services.push(child);
}
